src/backend/distributed/executor/executor.c

```diff
@@ -2022,6 +2117,20 @@ PlacementExecutionDone(TaskPlacementExecution *placementExecution, bool succeede
 	}
 	else
 	{
+		if (ShouldMarkPlacementsInvalidOnFailure(execution))
+		{
+			ShardPlacement *shardPlacement = placementExecution->shardPlacement;
+
+			/*
+			 * We only set shard state if its current state is FILE_FINALIZED, which
+			 * prevents overwriting shard state if it is already set at somewhere else.
+			 */
+			if (shardPlacement->shardState == FILE_FINALIZED)
```

Comment:
Minor unrelated note: it would be nice to rename FILE_FINALIZED and FILE_INACTIVE to something more  accurate (e.g. PLACEMENT_ACTIVE and PLACEMENT_INACTIVE). Probably not worth doing right now, but on a lazy day.